### PR TITLE
I164 processor allocator

### PR DESCRIPTION
**Pull request by gypsephi**
implemented processor allocator as mentioned in #164.

Free-invocation NYI due to issues with revokeOperation.
How to revoke the SC capability from users capmap and then push thread id to  free list of processor allocator?
Help needed:
https://github.com/ManyThreads/mythos/blob/e05070de4daf46633c7296172ba9aba5a2381d2c/kernel/objects/processor-allocator/objects/ProcessorAllocator.cc#L138
https://github.com/ManyThreads/mythos/blob/e05070de4daf46633c7296172ba9aba5a2381d2c/kernel/objects/processor-allocator/objects/ProcessorAllocator.cc#L152

No warnings.
Tested with qemu and IHK.

### CODE REVIEW

File: kernel/objects/scheduling-context/objects/SchedulingContext.cc [hunk 1]
```diff
@@ -41,9 +41,15 @@ namespace mythos {
     void SchedulingContext::unbind(handle_t* ec)
     {
         ASSERT(ec != nullptr);
-        MLOG_DETAIL(mlog::sched, "unbind", DVAR(ec->get()));
+        MLOG_INFO(mlog::sched, "unbind", DVAR(ec->get()));
         readyQueue.remove(ec);
         current_handle.store(nullptr);
+        if(readyQueue.empty()){
+          MLOG_DETAIL(mlog::sched, "call freeSC");
+          pluginProcessorAllocator.pa.freeSC(&paTask, home->getThreadID());
```
Comment (gypsephi) [inline kernel/objects/scheduling-context/objects/SchedulingContext.cc:49]: use a stub/event hook to make processor allocator optional?

Comment (kubanrob) [reply kernel/objects/scheduling-context/objects/SchedulingContext.cc:49]: That seems to be a good idea.

File: kernel/runtime/cxx/runtime/cxxsupport.cc [hunk 2]
```diff
@@ -49,23 +49,66 @@
...
+  , mythos::init::SIZE-mythos::init::APP_CAP_START> capAlloc;
+extern mythos::ProcessorAllocator pa;
 
 #ifndef NUM_CPUS
 #define NUM_CPUS (2)
 #endif
 
 #define PS_PER_TSC (0x0000000000000181)
 
+struct PthreadCleanerSemaphore{
```
Comment (kubanrob) [inline kernel/runtime/cxx/runtime/cxxsupport.cc:73]: This does not seem to be a semaphore to me.

File: kernel/runtime/cxx/runtime/cxxsupport.cc [hunk 3]
```diff
@@ -49,23 +49,66 @@
...
+    //MLOG_DETAIL(mlog::app, "PthreadCleaner exit", DVARhex(this), DVARhex(pthread_self()));
+    auto ec = flag.exchange(EXITED);
+    if(ec != FREE){
+      ASSERT(ec!=UNKNOWN);
+      mythos::syscall_signal(ec);
+    }
+  }
+
+  void wait(pthread_t t){
+    auto pcs = reinterpret_cast<PthreadCleanerSemaphore* >(t - (pthread_self() - reinterpret_cast<uintptr_t>(this)));
```
Comment (kubanrob) [inline kernel/runtime/cxx/runtime/cxxsupport.cc:96]: What is happening here? This definitely needs some sort of comment!

File: kernel/runtime/cxx/runtime/cxxsupport.cc [hunk 4]
```diff
@@ -49,23 +49,66 @@
...
+  void exit(){
+    //MLOG_DETAIL(mlog::app, "PthreadCleaner exit", DVARhex(this), DVARhex(pthread_self()));
+    auto ec = flag.exchange(EXITED);
+    if(ec != FREE){
+      ASSERT(ec!=UNKNOWN);
+      mythos::syscall_signal(ec);
+    }
+  }
+
+  void wait(pthread_t t){
```
Comment (kubanrob) [inline kernel/runtime/cxx/runtime/cxxsupport.cc:95]: Am I waiting for thread `t` with this function?

File: kernel/runtime/cxx/runtime/cxxsupport.cc [hunk 5]
```diff
@@ -49,23 +49,66 @@
...
+    auto pcs = reinterpret_cast<PthreadCleanerSemaphore* >(t - (pthread_self() - reinterpret_cast<uintptr_t>(this)));
+    //MLOG_DETAIL(mlog::app, "PthreadCleaner wait", DVARhex(pcs), DVARhex(this), DVARhex(pthread_self()), DVARhex(t));
+    while(pcs->flag.load() != EXITED){
+      mythos::CapPtr exp = FREE;
+      if(pcs->flag.compare_exchange_weak(exp, mythos_get_pthread_ec_self())){
+        //MLOG_DETAIL(mlog::app, "PthreadCleaner going to wait");
+        mythos_wait();
+      }
+    }
```
Comment (kubanrob) [inline kernel/runtime/cxx/runtime/cxxsupport.cc:104]: This function is supposed to be called by only one thread, right? It seems like this does busy waiting if another thread comes first.

File: kernel/runtime/cxx/runtime/cxxsupport.cc [hunk 6]
```diff
@@ -244,7 +288,8 @@ extern "C" int munmap(void *start, size_t len)
 extern "C" int unmapself(void *start, size_t len)
 {
     // dummy implementation
-    MLOG_DETAIL(mlog::app, "unmapself");
+    MLOG_ERROR(mlog::app, "unmapself: NYI!");
+    ASSERT(0);
```
Comment (kubanrob) [inline kernel/runtime/cxx/runtime/cxxsupport.cc:292]: `PANIC_MSG(false, "mlog::app, "unmapself: NYI!")`?

File: kernel/objects/execution-context/objects/ExecutionContext.cc [hunk 7]
```diff
@@ -55,7 +55,7 @@ namespace mythos {
                 DVARhex(flags.load()), isReady());
     if (needPreemption(prev) && !isReady()) {
         auto place = currentPlace.load();
-        ASSERT(place != nullptr);
+        //ASSERT(place != nullptr);
```
[the remark this answers is not in the document]
Comment (gypsephi) [reply kernel/objects/execution-context/objects/ExecutionContext.cc:58]: removed

File: kernel/objects/execution-context/objects/ExecutionContext.cc [hunk 8]
```diff
@@ -597,6 +597,7 @@ namespace mythos {
             auto place = currentPlace.load();
             if (place) synchronousAt(place) << [this]() {
                 this->saveState();
+                this->_sched.reset();
```
[the remark this answers is not in the document]
Comment (gypsephi) [reply kernel/objects/execution-context/objects/ExecutionContext.cc:600]: added comment

File: kernel/objects/scheduling-context/objects/SchedulingContext.cc (hunk 1 above)
Comment (gypsephi) [reply kernel/objects/scheduling-context/objects/SchedulingContext.cc]: done

File: kernel/objects/processor-allocator/objects/ProcessorAllocator.cc [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+
+  void ProcessorAllocator::invoke(Tasklet* t, Cap self, IInvocation* msg)
+  {
+    MLOG_DETAIL(mlog::pm, __func__, DVAR(t), DVAR(msg));
+    monitor.request(t, [=](Tasklet* t){
+        Error err = Error::NOT_IMPLEMENTED;
+        switch (msg->getProtocol()) {
+        //case protocol::KernelObject::proto:
+          //err = protocol::KernelObject::dispatchRequest(this, msg->getMethod(), self, msg);
+          //break;
```
[the remark this answers is not in the document]
Comment (gypsephi) [reply kernel/objects/processor-allocator/objects/ProcessorAllocator.cc:53]: removed

File: kernel/objects/processor-allocator/objects/ProcessorAllocator.cc [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+    auto id = alloc();
+
+    if(id){
+      MLOG_DETAIL(mlog::pm, "allocated ", DVAR(*id));
+
+      auto data = msg->getMessage()->read<protocol::ProcessorAllocator::Alloc>();
+
+      optional<CapEntry*> dstEntry;
+      if(data.dstSpace() == null_cap){ // direct access
+        //dstEntry = msg->lookupEntry(data.dstPtr, 32, true); // lookup for write access
```
[the remark this answers is not in the document]
Comment (gypsephi) [reply kernel/objects/processor-allocator/objects/ProcessorAllocator.cc:94]: removed

File: kernel/objects/processor-allocator/objects/ProcessorAllocator.cc [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+  }
+
+  Error ProcessorAllocator::invokeFree(Tasklet* /*t*/, Cap, IInvocation* /*msg*/){
+    MLOG_ERROR(mlog::pm, __func__, ": NYI!");  
+    //auto data = msg->getMessage()->cast<protocol::ProcessorAllocator::Free>();
+    //ASSERT(data->sc() >= init::SCHEDULERS_START);
+    //cpu::ThreadID id = data->sc() - init::SCHEDULERS_START;
+    //ASSERT(id < cpu::getNumThreads());
+    //MLOG_ERROR(mlog::pm, "free SC", DVAR(data->sc()), DVAR(id));
+    //revokeOp._revoke(t, this, sc[id], this);
```
[the remark this answers is not in the document]
Comment (gypsephi) [reply kernel/objects/processor-allocator/objects/ProcessorAllocator.cc:145]: lets talk, i think we need a new revocation mechanism...

File: kernel/runtime/cxx/runtime/cxxsupport.cc (hunk 2 above)
Comment (gypsephi) [reply kernel/runtime/cxx/runtime/cxxsupport.cc:73]: renamed

File: kernel/runtime/cxx/runtime/cxxsupport.cc (hunk 3 above)
Comment (gypsephi) [reply kernel/runtime/cxx/runtime/cxxsupport.cc]: added comments

File: kernel/runtime/cxx/runtime/cxxsupport.cc (hunk 4 above)
Comment (gypsephi) [reply kernel/runtime/cxx/runtime/cxxsupport.cc:102]: added comment

File: kernel/runtime/cxx/runtime/cxxsupport.cc (hunk 5 above)
Comment (gypsephi) [reply kernel/runtime/cxx/runtime/cxxsupport.cc]: it is some kind of simple futex. try to register as waiter and then going to sleep (in a loop). added comments

File: kernel/runtime/cxx/runtime/cxxsupport.cc (hunk 6 above)
Comment (gypsephi) [reply kernel/runtime/cxx/runtime/cxxsupport.cc]: done

File: kernel/objects/processor-allocator/objects/ProcessorAllocator.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (gypsephi) [reply kernel/objects/processor-allocator/objects/ProcessorAllocator.cc:145]: the invokeFree method stays unimplemented until a suitable revoke operation gets implemented.
See #189

File: kernel/objects/processor-allocator/objects/ProcessorAllocator.cc [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+    //free(id);
+    //return Error::SUCCESS;
+    return Error::NOT_IMPLEMENTED;
+  }
+
+  void ProcessorAllocator::freeSC(Tasklet* t, cpu::ThreadID id){
+    MLOG_DETAIL(mlog::pm, "freeSC", DVAR(id));
+    monitor.request(t, [=](Tasklet* t){
+      MLOG_DETAIL(mlog::pm, "monitor free", DVAR(id));
+      //todo: use IResult to call free(id) after cap is revoked
```
[the remark this answers is not in the document]
Comment (gypsephi) [reply kernel/objects/processor-allocator/objects/ProcessorAllocator.cc]: fixed using resonse method of IResult interface.

**Review by kubanrob — APPROVED** [review]

**Pull request merged**